api/controller.go

```diff
@@ -2372,6 +2396,68 @@ func (c *Controller) DetachPolicyFromGroupHandler() authop.DetachPolicyFromGroup
 	})
 }
 
+func (c *Controller) RefsRefsRestoreHandler() refs.RefsRestoreHandler {
+	return refs.RefsRestoreHandlerFunc(func(params refs.RefsRestoreParams, user *models.User) middleware.Responder {
+		deps, err := c.setupRequest(user, params.HTTPRequest, []permissions.Permission{
+			{
+				Action:   permissions.ListTagsAction,
+				Resource: permissions.RepoArn(params.Repository),
+			},
+			{
+				Action:   permissions.ListBranchesAction,
+				Resource: permissions.RepoArn(params.Repository),
+			},
+			{
+				Action:   permissions.ListCommitsAction,
+				Resource: permissions.RepoArn(params.Repository),
+			},
+		})
+		if err != nil {
+			return refs.NewRefsRestoreUnauthorized().
+				WithPayload(responseErrorFrom(err))
+		}
+		deps.LogAction("restore_repository_refs")
+
+		repo, err := deps.Cataloger.GetRepository(deps.ctx, params.Repository)
+		if errors.Is(err, catalog.ErrRepositoryNotFound) {
+			return refs.NewRefsRestoreNotFound().
+				WithPayload(responseErrorFrom(err))
+		} else if err != nil {
+			return refs.NewRefsRestoreDefault(http.StatusInternalServerError).
+				WithPayload(responseErrorFrom(err))
+		}
+
+		// ensure no refs currently found
+		_, _, err = deps.Cataloger.ListCommits(deps.ctx, repo.Name, repo.DefaultBranch, "", 1)
+		if !errors.Is(err, graveler.ErrNotFound) {
+			return refs.NewRefsRestoreDefault(http.StatusInternalServerError).
+				WithPayload(responseError("restoring refs is supported only for bare repositories"))
+		}
+
+		// load commits
+		err = deps.Cataloger.LoadCommits(deps.ctx, repo.Name, params.Manifest.CommitsMetaRangeID)
+		if err != nil {
+			return refs.NewRefsRestoreDefault(http.StatusInternalServerError).
+				WithPayload(responseErrorFrom(err))
+		}
+
+		err = deps.Cataloger.LoadBranches(deps.ctx, repo.Name, params.Manifest.BranchesMetaRangeID)
+		if err != nil {
+			return refs.NewRefsRestoreDefault(http.StatusInternalServerError).
+				WithPayload(responseErrorFrom(err))
+		}
+
+		err = deps.Cataloger.LoadTags(deps.ctx, repo.Name, params.Manifest.TagsMetaRangeID)
+		if err != nil {
+			return refs.NewRefsRestoreDefault(http.StatusInternalServerError).
+				WithPayload(responseErrorFrom(err))
+		}
+
```

Comment:
If we failed before this time, we are in an inconsistent state.  I understand why, but I am not sure how the user could clean up the mess we leave in order to be able to try again.

Reply:
Since ref-loading only works on an initial bare repo, simplest thing to do would be to drop the repo and start over on failure.
All actions should in theory be idempotent, so it could be possible to allow the user to retry but I'm not sure it's worth it

Reply:
Cool.  Maybe just make the error message explain this here, then.